src/CalcManager/UnitConverter.cpp

```diff
@@ -842,100 +844,59 @@ void UnitConverter::Calculate()
-    if (conversionTable[m_toType].ratio == 1.0 && conversionTable[m_toType].offset == 0.0)
+    if (AnyUnitIsEmpty() || (conversionTable[m_toType].ratio == 1.0 && conversionTable[m_toType].offset == 0.0))
     {
         m_returnDisplay = m_currentDisplay;
         m_returnHasDecimal = m_currentHasDecimal;
-        TrimString(m_returnDisplay);
+        TrimTrailingZeros(m_returnDisplay);
     }
     else
     {
-        returnValue = Convert(returnValue, conversionTable[m_toType]);
-        m_returnDisplay = RoundSignificant(returnValue, MAXIMUMDIGITSALLOWED);
-        TrimString(m_returnDisplay);
-        int numPreDecimal = (int)m_returnDisplay.size();
-        if (m_returnDisplay.find(L'.') != m_returnDisplay.npos)
-        {
-            numPreDecimal = (int)m_returnDisplay.find(L'.');
-        }
-        if (returnValue < 0)
-        {
-            numPreDecimal--;
-        }
+        double currentValue = stod(m_currentDisplay);
+        double returnValue = Convert(currentValue, conversionTable[m_toType]);
 
-        if (numPreDecimal > MAXIMUMDIGITSALLOWED || (returnValue != 0 && abs(returnValue) < MINIMUMDECIMALALLOWED))
+        auto isCurrencyConverter = m_currencyDataLoader != nullptr && m_currencyDataLoader->SupportsCategory(this->m_currentCategory);
+        if (isCurrencyConverter)
         {
-            wstringstream out(wstringstream::out);
-            out << scientific << returnValue;
-            m_returnDisplay = out.str();
+            // We don't need to trim the value when it's a currency.
+            m_returnDisplay = RoundSignificantDigits(returnValue, MAXIMUMDIGITSALLOWED);
+            TrimTrailingZeros(m_returnDisplay);
         }
         else
         {
-            returnValue = stod(m_returnDisplay);
-            wstring returnString;
-            if (m_currentDisplay.size() <= OPTIMALDIGITSALLOWED && abs(returnValue) >= OPTIMALDECIMALALLOWED)
+            int numPreDecimal = GetNumberDigitsWholeNumberPart(returnValue);
+            if (numPreDecimal > MAXIMUMDIGITSALLOWED || (returnValue != 0 && abs(returnValue) < MINIMUMDECIMALALLOWED))
             {
-                returnString = RoundSignificant(returnValue, OPTIMALDIGITSALLOWED - min(numPreDecimal, OPTIMALDIGITSALLOWED));
+                m_returnDisplay = ToScientificNumber(returnValue);
             }
             else
             {
-                returnString = RoundSignificant(returnValue, MAXIMUMDIGITSALLOWED - min(numPreDecimal, MAXIMUMDIGITSALLOWED));
+                int currentNumberSignificantDigits = GetNumberDigits(m_currentDisplay);
+                int precision = 0;
+                if (abs(returnValue) < OPTIMALDECIMALALLOWED)
+                {
+                    precision = MAXIMUMDIGITSALLOWED;
+                }
+                else
+                {
+                    precision = max(0, max(OPTIMALDIGITSALLOWED, min(MAXIMUMDIGITSALLOWED, currentNumberSignificantDigits)) - numPreDecimal);
```

Comment:
I think this makes sense, can we add a small comment giving an explanation for the logic here?